flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReaderBase.java

```diff
@@ -0,0 +1,187 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.connector.source.lib.util;
+
+import org.apache.flink.annotation.Public;
+import org.apache.flink.api.connector.source.ReaderOutput;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.core.io.InputStatus;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A {@link SourceReader} that returns the values of an iterator, supplied via an {@link
+ * IteratorSourceSplit}.
+ *
+ * <p>The {@code IteratorSourceSplit} is also responsible for taking the current iterator and
+ * turning it back into a split for checkpointing.
+ *
+ * @param <E> The type of events returned by the reader.
+ * @param <IterT> The type of the iterator that produces the events. This type exists to make the
+ *     conversion between iterator and {@code IteratorSourceSplit} type safe.
+ * @param <SplitT> The concrete type of the {@code IteratorSourceSplit} that creates and converts
+ *     the iterator that produces this reader's elements.
+ */
+@Public
+public abstract class IteratorSourceReaderBase<
+                E, O, IterT extends Iterator<E>, SplitT extends IteratorSourceSplit<E, IterT>>
+        implements SourceReader<O, SplitT> {
+
+    /** The context for this reader, to communicate with the enumerator. */
+    private final SourceReaderContext context;
+
+    /** The availability future. This reader is available as soon as a split is assigned. */
+    private CompletableFuture<Void> availability;
+
+    /**
+     * The iterator producing data. Non-null after a split has been assigned. This field is null or
+     * non-null always together with the {@link #currentSplit} field.
+     */
+    @Nullable private IterT iterator;
+
+    /**
+     * The split whose data we return. Non-null after a split has been assigned. This field is null
+     * or non-null always together with the {@link #iterator} field.
+     */
+    @Nullable private SplitT currentSplit;
+
+    /** The remaining splits that were assigned but not yet processed. */
+    private final Queue<SplitT> remainingSplits;
+
+    private boolean noMoreSplits;
+
+    public IteratorSourceReaderBase(SourceReaderContext context) {
+        this.context = checkNotNull(context);
+        this.availability = new CompletableFuture<>();
+        this.remainingSplits = new ArrayDeque<>();
+    }
+
+    // ------------------------------------------------------------------------
+
+    @Override
+    public final void start() {
```

Comment:
I will remove the `final` flag because it's not source-compatible.

Reply:
This not being detected is FYI another bug that we found.

Reply:
Are there any actions required from my side?

Reply:
No :)